### PR TITLE
etcdutils(engine): refine etcd join test and prepare etcd lib

### LINKED ISSUE #5782: engine flaking test TestLeaderLoopSuccess
### Which jobs are flaking?

TestLeaderLoopSuccess 

### Which test(s) are flaking?

engine-UT

```
[2022-06-08T03:42:41.203Z] === FAIL: engine/servermaster TestLeaderLoopSuccess (0.00s)

[2022-06-08T03:42:41.203Z]     etcd_utils.go:59: 

[2022-06-08T03:42:41.203Z]         	Error Trace:	etcd_utils.go:59

[2022-06-08T03:42:41.203Z]         	            				campaign_test.go:50

[2022-06-08T03:42:41.203Z]         	Error:      	Expected nil, but got: [DFLOW:ErrMasterStartEmbedEtcdFail]failed to start embed etcd: listen tcp 127.0.0.1:36440: bind: address already in use

[2022-06-08T03:42:41.203Z]         	Test:       	TestLeaderLoopSuccess
```

### Jenkins logs or GitHub Actions link

https://ci.pingcap.net/blue/organizations/jenkins/ut-check/detail/ut-check/16104/pipeline/

### Anything else we need to know

- Does this test exist for other branches as well?

- Has there been a high frequency of failure lately?

### PR DESCRIPTION
**Pull request by amyangfei**
### What problem does this PR solve?

Issue Number: close #5782

### What is changed and how it works?

- Fix wrong logic in the last step of etcd join test
- Add retry to `test.PrepareEtcd` function

#### Tests 

 - Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by amyangfei** [comment]
/run-
check-issue-triage-complete

**Review by maxshuang — APPROVED** [review]

File: engine/test/etcd_utils.go [hunk 1]
```diff
@@ -39,12 +40,39 @@ func allocTempURL(t *testing.T) string {
...
+	err := retry.Do(context.TODO(), func() error {
+		var err error
+		advertiseAddr, etcd, client, cleanFn, err = prepareEtcdOnce(t, name)
+		return err
+	},
+		retry.WithBackoffBaseDelay(1000 /* 1000 ms */),
+		retry.WithBackoffMaxDelay(3000 /* 3 seconds */),
+		retry.WithMaxTries(3 /* fail after 10 seconds*/),
+		retry.WithIsRetryableErr(func(err error) bool {
+			if strings.Contains(err.Error(), "address already in use") {
```
Comment (CharlesCheung96) [inline engine/test/etcd_utils.go:55]: Why retry only in this condition?

Comment (amyangfei) [reply engine/test/etcd_utils.go:55]: Other errors are not expected.

Comment (amyangfei) [reply engine/test/etcd_utils.go:55]: ptal @CharlesCheung96

**Pull request merged**